sql/core/src/main/scala/org/apache/spark/sql/execution/joins/JoinCodegenSupport.scala

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.joins
+
+import org.apache.spark.sql.catalyst.expressions.{BindReferences, BoundReference}
+import org.apache.spark.sql.catalyst.expressions.codegen._
+import org.apache.spark.sql.catalyst.expressions.codegen.Block._
+import org.apache.spark.sql.catalyst.plans.InnerLike
+import org.apache.spark.sql.execution.{CodegenSupport, SparkPlan}
+
+/**
+ * An interface for those join physical operators that support codegen.
+ */
+trait JoinCodegenSupport extends CodegenSupport with BaseJoinExec {
+
+  /**
+   * Generate the (non-equi) condition used to filter joined rows. This is used in Inner, Left Semi
+   * and Left Anti joins.
+   */
+  protected def getJoinCondition(
+      ctx: CodegenContext,
+      input: Seq[ExprCode],
+      streamedPlan: SparkPlan,
+      buildPlan: SparkPlan): (String, String, Seq[ExprCode]) = {
+    val buildRow = ctx.freshName("buildRow")
+    val buildVars = genBuildSideVars(ctx, buildRow, buildPlan)
+    val checkCondition = if (condition.isDefined) {
+      val expr = condition.get
+      // evaluate the variables from build side that used by condition
```

Comment:
This is coped so my question may not be related to this PR: the join condition can refer to columns from stream side, why don't we need to evaluate variables from the stream side?

Reply:
This seems like a problem. In `HashJoin`, we have this at the beginning
```
         |// generate join key for stream side
         |${keyEv.code}
```

Shall we do something similar in BroadcastNestedLoopJoinExec?

Reply:
@cloud-fan - `input` here is the generated code for streamed side. The code is generated below with `ctx.currentVars = input ++ buildVars`.

Reply:
> Shall we do something similar in BroadcastNestedLoopJoinExec?

We don't have join key for streamed side in `BroadcastNestedLoopJoinExec`, i.e. non-equal joins.

Reply:
`input` also includes `buidVars`, my question is why this line of code here only handles build side not stream side.

Reply:
`evaluateRequiredVariables` is for performance purpose (eagerly evaluate some expressios and put them in vars), so this is not about correctness and is hard to test.

Reply:
The variables from stream side are be already evaluated before `consume` is called. So you don't need to include stream side variables again.

Reply:
@viirya hmmm are you sure about it? AFAIK for operators like `Filter`, it doesn't evaluate all the child outputs, only the ones required by the condition, though `Filter.output` is `child.output`.

Reply:
For Filter or Project, we do some evaluation deferring. Normally `CodegenSupport.usedInputs` is set to  `references`. `HashJoin` doesn't override it.

Reply:
After a second check, I think we don't need to evaluate variable from stream side here, as `BroadcastNestedLoopJoinExec`'s stream child node should already have evaluated them in [consume()](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala#L177) before calling `BroadcastNestedLoopJoinExec.doConsume()`. As @viirya pointed out, given `BroadcastNestedLoopJoinExec.usedInputs()` is `references`, it should contain all attributes of `condition` from stream side.

Thanks @viirya for pointing this out and careful review. @cloud-fan wondering what do you think?

Reply:
ah, good catch! @viirya is right, we don't need to do that.

Reply:
Updated to evaluate build side only. Same as before in `HashJoin.scala`.